notes/error.md

```diff
@@ -1,349 +1,444 @@
+
+$$
+\text{Relative Error} = \frac{|x - \hat{x}|}{|x|} \geq 10^{-n + 1}
+$$
+
+then $$\hat{x}$$ has **at most** $$n$$ accurate significant digits. 
+
+In general, we will use the rule-of-thumb for calculating an upper bound of the relative error: if an approximation has $$n$$ accurate significant digits, then the relative error is
+
+$$
+\frac{|x - \hat{x}|}{|x|} \leq 10^{-n+1}
+$$
+
+
+
+## Absolute and Relative Error of Vectors
+If our calculated quantities are vectors then instead of using the absolute
+value function, we can use the norm instead. Thus, our formulas become
+
+$$
+\begin{equation}
+\text{Absolute Error} = \|\mathbf{x} - \mathbf{\hat{x}}\|
+\end{equation}
+$$
+
+$$
+\begin{equation}
+\text{Relative Error} = \frac{\|\mathbf{x} - \mathbf{\hat{x}}\|}{\|\mathbf{x}\|}
+\end{equation}
+$$
+
+
+We take the norm of the difference (and not the difference of the norms),
+because we are interested in how far apart these two quantities are.
+This formula is similar to finding that difference then using the vector
+norm to find the length of that difference vector.
+
+
+## Truncation Error vs. Rounding Error
+
+**Rounding error** is the error that occurs from rounding values in a computation. This occurs constantly since computers use finite precision and have a limit on the memory available for storing one numerical value. Approximating $$\frac{1}{3} = 0.33333\dots$$ with a finite decimal expansion is an example of rounding error.
+
+
+**Truncation error** is the error from using an approximate algorithm in place of an exact mathematical procedure or function. For example, in the case of evaluating functions, we may represent our function by a finite Taylor series up to degree $$n$$. The truncation error is the error that is incurred by not using the $$n+1$$ term and above.
+
+
+## Big-O Notation
+
+Big-O notation is used to understand and describe asymptotic behavior.
+The definition in the cases of approching 0 or $$\infty$$ are as follows:
+
+
+Let $$f$$ and $$g$$ be two functions. Then
+$$f(x) = \mathcal{O}(g(x))$$ as $$x \rightarrow \infty$$
+if and only if there exists a value $$M$$ and some $$x_0$$ such that
+$$|f(x)| \leq M|g(x)|$$ $$\forall x$$ where $$x\geq x_0$$
+
+
+Let $$f$$ and $$g$$ be two functions. Then
+$$f(h) = \mathcal{O}(g(h))$$ as $$h \rightarrow 0$$
+if and only if there exists a value $$M$$ and some $$h_0$$ such that
+$$|f(h)| \leq M|g(h)|$$ $$\forall h$$ where $$0 < h < h_0$$
+
+
+But what if we want to consider the function approaching an arbitrary value?
+Then we can redefine the expression as:
+
+
+Let $$f$$ and $$g$$ be two functions. Then
+$$f(x) = \mathcal{O}(g(x))$$ as $$x \rightarrow a$$
+if and only if there exists a value $$M$$ and some $$\delta$$ such that
+$$|f(x)| \leq M|g(x)|$$ $$\forall x$$ where $$0 < |x − a| < \delta$$
+
+For example, consider the function $$f(x) = 2x^2 + 27x + 1000$$
+
+When $$x \rightarrow \infty$$, the term $$x^2$$ is the most significant, and hence $$f(x) = O(x^2)$$
+

```

Comment:
do we need a big O hierarchy chart here (ie n^n > n! > 2^n > n^p > n lg n > n > lg n > 1, etc)?